File: clients/python/anna/bench.py
"""Benchmark infrastructure for measuring KVS throughput and latency."""

import time as time_mod
from .lattices import LWWPairLattice


def bench_key(n):
    """Format a key index as a zero-padded 8-character string."""
    return str(n).zfill(8)


def run_bench(client, num_keys=1000, value_size=256, duration=10,
              report_period=2, workloads=None):
    """Run benchmarks against the KVS.

    Args:
        client: An AnnaTcpClient instance.
        num_keys: Number of keys in the key space.
        value_size: Size of values in bytes.
        duration: Duration of each workload in seconds.
        report_period: Seconds between throughput reports.
        workloads: List of workload names (GET, PUT, MIXED).

    Returns:
        List of (workload, avg_throughput, avg_latency_us, total_ops, elapsed) tuples.
    """
    if num_keys <= 0:
        raise ValueError("num_keys must be > 0")
    if duration <= 0:
        raise ValueError("duration must be > 0")
    if report_period <= 0:
        raise ValueError("report_period must be > 0")

    if workloads is None:
        workloads = ["GET", "PUT", "MIXED"]

    value = "a" * value_size
    results = []

    # Validate all workloads upfront.
    for wl in workloads:
        if wl not in ("GET", "PUT", "MIXED"):
            raise ValueError(f"Invalid workload: {wl}. Must be GET, PUT, or MIXED.")

    # Warmup once, shared across all workloads.
    print(f"Warming up {num_keys} keys ({value_size} bytes each)...")
    warmup_start = time_mod.monotonic()
    for i in range(1, num_keys + 1):
        ts = time_mod.time_ns()
        client.put(bench_key(i), LWWPairLattice(ts, value.encode()))
    warmup_ms = (time_mod.monotonic() - warmup_start) * 1000
    print(f"Warmup complete in {warmup_ms:.0f} ms")

    for wl in workloads:
        print(f"Running {wl} benchmark for {duration}s "
              f"({num_keys} keys, {value_size} B values)...")

        total_ops = 0
        epoch_ops = 0
        seed = int(time_mod.monotonic() * 1e9) & 0xFFFFFFFF

        bench_start = time_mod.monotonic()
        epoch_start = bench_start

        while True:
            seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
            k = (seed % num_keys) + 1
            key = bench_key(k)

            if wl == "GET":
                client.get(key)
                total_ops += 1
                epoch_ops += 1
            elif wl == "PUT":
                ts = time_mod.time_ns()
                client.put(key, LWWPairLattice(ts, value.encode()))
                total_ops += 1
                epoch_ops += 1
            else:  # MIXED
                ts = time_mod.time_ns()
                client.put(key, LWWPairLattice(ts, value.encode()))
                client.get(key)
                total_ops += 2
                epoch_ops += 2

            now = time_mod.monotonic()
            if now - epoch_start >= report_period:
                secs = now - epoch_start
                throughput = epoch_ops / secs
                print(f"[Epoch] Throughput: {int(throughput)} ops/sec")
                epoch_ops = 0
                epoch_start = now

            if now - bench_start >= duration:
                break

        elapsed = time_mod.monotonic() - bench_start
        avg_tp = total_ops / elapsed if elapsed > 0 else 0
        avg_lat = 1_000_000.0 / avg_tp if avg_tp > 0 else 0

        print(f"\n=== {wl} Results ===")
        print(f"Total ops:      {total_ops}")
        print(f"Elapsed:        {elapsed:.2f} s")
        print(f"Avg throughput: {int(avg_tp)} ops/sec")
        print(f"Avg latency:    {avg_lat:.1f} us/op")
        print()
        results.append((wl, avg_tp, avg_lat, total_ops, elapsed))

    print("\n=== Benchmark Summary (Python) ===")
    print(f"{'Workload':<10} {'Ops/sec':>12} {'Latency(us)':>14} "
          f"{'Total ops':>12} {'Time(s)':>10}")
    print("-" * 58)
    for wl, tp, lat, ops, secs in results:
        print(f"{wl:<10} {int(tp):>12} {lat:>14.1f} {ops:>12} {secs:>10.2f}")

    return results

File: clients/rust/src/lib/bench.rs
//! Benchmark infrastructure for measuring KVS throughput and latency.

use crate::errors::Result;
use std::time::{Duration, Instant};

/// Trait abstracting the KVS operations needed for benchmarking.
/// Implemented by `KVSClient` for production use, and by test mocks.
pub trait BenchClient {
    /// Put a key-value pair.
    fn put_val(
        &mut self,
        key: &str,
        value: &str,
    ) -> impl std::future::Future<Output = Result<()>> + Send;
    /// Get a key's value.
    fn get_val(&mut self, key: &str) -> impl std::future::Future<Output = Result<String>> + Send;
}

impl BenchClient for crate::kvs_client::KVSClient {
    async fn put_val(&mut self, key: &str, value: &str) -> Result<()> {
        self.put(key, value).await
    }
    async fn get_val(&mut self, key: &str) -> Result<String> {
        self.get(key).await
    }
}

/// Configuration for a benchmark run.
pub struct BenchConfig {
    /// Number of keys in the key space.
    pub num_keys: u64,
    /// Size of each value in bytes.
    pub value_size: usize,
    /// Total duration of each workload.
    pub duration: Duration,
    /// Seconds between throughput reports.
    pub report_period: Duration,
    /// Workload names to run (GET, PUT, MIXED).
    pub workloads: Vec<String>,
}

impl Default for BenchConfig {
    fn default() -> Self {
        Self {
            num_keys: 1000,
            value_size: 256,
            duration: Duration::from_secs(10),
            report_period: Duration::from_secs(2),
            workloads: vec!["GET".into(), "PUT".into(), "MIXED".into()],
        }
    }
}

/// Result of a single workload run.
pub struct WorkloadResult {
    /// Workload name (GET, PUT, or MIXED).
    pub name: String,
    /// Total number of operations completed.
    pub total_ops: u64,
    /// Wall-clock time elapsed.
    pub elapsed: Duration,
}

impl WorkloadResult {
    /// Average operations per second.
    pub fn ops_per_sec(&self) -> f64 {
        self.total_ops as f64 / self.elapsed.as_secs_f64()
    }

    /// Average microseconds per operation.
    pub fn us_per_op(&self) -> f64 {
        if self.total_ops == 0 {
            return 0.0;
        }
        self.elapsed.as_micros() as f64 / self.total_ops as f64
    }
}

/// Format a key index as a zero-padded 8-character string.
pub fn bench_key(index: u64) -> String {
    format!("{:08}", index)
}

/// Simple pseudo-random number generator (LCG).
pub struct SimpleRng {
    state: u64,
}

impl SimpleRng {
    /// Create a new RNG with the given seed.
    pub fn new(seed: u64) -> Self {
        Self { state: seed }
    }

    /// Returns a pseudo-random u64.
    pub fn next_u64(&mut self) -> u64 {
        self.state = self
            .state
            .wrapping_mul(6_364_136_223_846_793_005)
            .wrapping_add(1);
        self.state
    }

    /// Returns a value in `[0, bound)`.
    pub fn next_bounded(&mut self, bound: u64) -> u64 {
        self.next_u64() % bound
    }
}

/// Run the full benchmark: warmup, then each workload, then summary.
pub async fn run_bench(
    client: &mut (impl BenchClient + Send),
    config: &BenchConfig,
) -> Result<Vec<WorkloadResult>> {
    if config.num_keys == 0 {
        return Err(crate::errors::Error::Process(
            "num_keys must be > 0".to_string(),
        ));
    }
    let value: String = "a".repeat(config.value_size);

    println!(
        "Benchmark (Rust): keys={}, value_size={}, duration={}s, report={}s",
        config.num_keys,
        config.value_size,
        config.duration.as_secs(),
        config.report_period.as_secs(),
    );

    // Warmup
    println!("Warming up {} keys...", config.num_keys);
    let warmup_start = Instant::now();
    for i in 0..config.num_keys {
        client.put_val(&bench_key(i), &value).await?;
    }
    println!(
        "Warmup complete: {} keys in {:.2}s",
        config.num_keys,
        warmup_start.elapsed().as_secs_f64()
    );

    let mut results = Vec::new();
    for wl in &config.workloads {
        let result = run_workload(
            client,
            wl,
            config.num_keys,
            &value,
            config.duration,
            config.report_period,
        )
        .await?;
        results.push(result);
    }

    // Summary
    println!();
    println!("=== Benchmark Summary (Rust) ===");
    println!(
        "{:<10} {:>12} {:>12} {:>12} {:>10}",
        "Workload", "ops/sec", "us/op", "total_ops", "elapsed"
    );
    println!("{}", "-".repeat(60));
    for r in &results {
        println!(
            "{:<10} {:>12.1} {:>12.1} {:>12} {:>9.2}s",
            r.name,
            r.ops_per_sec(),
            r.us_per_op(),
            r.total_ops,
            r.elapsed.as_secs_f64()
        );
    }

    Ok(results)
}

async fn run_workload(
    client: &mut (impl BenchClient + Send),
    workload: &str,
    num_keys: u64,
    value: &str,
    duration: Duration,
    report_period: Duration,
) -> Result<WorkloadResult> {
    println!();
    println!("--- {} workload ---", workload);

    let seed = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos() as u64)
        .unwrap_or(42);
    let mut rng = SimpleRng::new(seed);

    let start = Instant::now();
    let mut total_ops: u64 = 0;
    let mut epoch_ops: u64 = 0;
    let mut last_report = start;

    while start.elapsed() < duration {
        let key = bench_key(rng.next_bounded(num_keys));

        match workload {
            "GET" => {
                client.get_val(&key).await?;
                total_ops += 1;
                epoch_ops += 1;
            }
            "PUT" => {
                client.put_val(&key, value).await?;
                total_ops += 1;
                epoch_ops += 1;
            }
            "MIXED" => {
                client.put_val(&key, value).await?;
                client.get_val(&key).await?;
                total_ops += 2;
                epoch_ops += 2;
            }
            _ => {
                return Err(crate::errors::Error::Process(format!(
                    "Invalid workload: {}",
                    workload
                )))
            }
        }

        let now = Instant::now();
        if now.duration_since(last_report) >= report_period {
            let secs = now.duration_since(last_report).as_secs_f64();
            println!(
                "  [{:>6.1}s] {:>10.1} ops/sec  ({} ops in {:.2}s)",
                now.duration_since(start).as_secs_f64(),
                epoch_ops as f64 / secs,
                epoch_ops,
                secs,
            );
            epoch_ops = 0;
            last_report = now;
        }
    }

    let elapsed = start.elapsed();
    println!(
        "{} complete: {} ops in {:.2}s ({:.1} ops/sec)",
        workload,
        total_ops,
        elapsed.as_secs_f64(),
        total_ops as f64 / elapsed.as_secs_f64()
    );

    Ok(WorkloadResult {
        name: workload.to_string(),
        total_ops,
        elapsed,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bench_key_pads_small_numbers() {
        assert_eq!(bench_key(1), "00000001");
        assert_eq!(bench_key(42), "00000042");
        assert_eq!(bench_key(1000000), "01000000");
    }

    #[test]
    fn bench_key_large_numbers() {
        assert_eq!(bench_key(99999999), "99999999");
        assert_eq!(bench_key(100000000), "100000000");
    }

    #[test]
    fn simple_rng_bounded() {
        let mut rng = SimpleRng::new(12345);
        for _ in 0..100 {
            assert!(rng.next_bounded(1000) < 1000);
        }
    }

    #[test]
    fn simple_rng_different_seeds() {
        let mut rng1 = SimpleRng::new(1);
        let mut rng2 = SimpleRng::new(2);
        let mut differ = false;
        for _ in 0..10 {
            if rng1.next_bounded(1000) != rng2.next_bounded(1000) {
                differ = true;
                break;
            }
        }
        assert!(differ);
    }

    #[test]
    fn workload_result_ops_per_sec() {
        let r = WorkloadResult {
            name: "GET".into(),
            total_ops: 1000,
            elapsed: Duration::from_secs(2),
        };
        assert!((r.ops_per_sec() - 500.0).abs() < 0.1);
        assert!((r.us_per_op() - 2000.0).abs() < 0.1);
    }

    #[test]
    fn workload_result_zero_ops() {
        let r = WorkloadResult {
            name: "PUT".into(),
            total_ops: 0,
            elapsed: Duration::from_secs(1),
        };
        assert_eq!(r.us_per_op(), 0.0);
    }

    #[test]
    fn default_bench_config() {
        let cfg = BenchConfig::default();
        assert_eq!(cfg.num_keys, 1000);
        assert_eq!(cfg.value_size, 256);
        assert_eq!(cfg.duration, Duration::from_secs(10));
        assert_eq!(cfg.workloads, vec!["GET", "PUT", "MIXED"]);
    }

    /// Mock client for testing bench functions without a real server.
    struct MockBenchClient;

    impl BenchClient for MockBenchClient {
        async fn put_val(&mut self, _key: &str, _value: &str) -> Result<()> {
            Ok(())
        }
        async fn get_val(&mut self, _key: &str) -> Result<String> {
            Ok("mock_value".to_string())
        }
    }

    #[tokio::test]
    async fn run_bench_get_workload() {
        let mut client = MockBenchClient;
        let config = BenchConfig {
            num_keys: 5,
            value_size: 16,
            duration: Duration::from_secs(1),
            report_period: Duration::from_secs(1),
            workloads: vec!["GET".into()],
        };
        let results = run_bench(&mut client, &config)
            .await
            .expect("bench should succeed");
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].name, "GET");
        assert!(results[0].total_ops > 0);
    }

    #[tokio::test]
    async fn run_bench_put_workload() {
        let mut client = MockBenchClient;
        let config = BenchConfig {
            num_keys: 5,
            value_size: 16,
            duration: Duration::from_secs(1),
            report_period: Duration::from_secs(1),
            workloads: vec!["PUT".into()],
        };
        let results = run_bench(&mut client, &config)
            .await
            .expect("bench should succeed");
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].name, "PUT");
        assert!(results[0].total_ops > 0);
    }

    #[tokio::test]
    async fn run_bench_mixed_workload() {
        let mut client = MockBenchClient;
        let config = BenchConfig {
            num_keys: 5,
            value_size: 16,
            duration: Duration::from_secs(1),
            report_period: Duration::from_secs(1),
            workloads: vec!["MIXED".into()],
        };
        let results = run_bench(&mut client, &config)
            .await
            .expect("bench should succeed");
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].name, "MIXED");
        assert!(results[0].total_ops > 0);
    }

    #[tokio::test]
    async fn run_bench_all_workloads() {
        let mut client = MockBenchClient;
        let config = BenchConfig {
            num_keys: 5,
            value_size: 16,
            duration: Duration::from_secs(1),
            report_period: Duration::from_secs(1),
            workloads: vec!["GET".into(), "PUT".into(), "MIXED".into()],
        };
        let results = run_bench(&mut client, &config)
            .await
            .expect("bench should succeed");
        assert_eq!(results.len(), 3);
        assert_eq!(results[0].name, "GET");
        assert_eq!(results[1].name, "PUT");
        assert_eq!(results[2].name, "MIXED");
    }

    #[tokio::test]
    async fn run_bench_zero_keys_fails() {
        let mut client = MockBenchClient;
        let config = BenchConfig {
            num_keys: 0,
            value_size: 16,
            duration: Duration::from_secs(1),
            report_period: Duration::from_secs(1),
            workloads: vec!["GET".into()],
        };
        assert!(run_bench(&mut client, &config).await.is_err());
    }
}
